### PR TITLE
Remove split CCL

### PR DESCRIPTION
**Pull request by WeiyiVirtonomy**
The split cell linked list is replaced by a function called `particle_for_split`, where the cells are accessed by a stride of 3 in the indices. 

TODO:
Add sequential version
Add unit test

### CODE REVIEW

File: src/for_2D_build/meshes/mesh_iterators.hpp [hunk 1]
```diff
@@ -84,4 +84,46 @@ void mesh_parallel_for(const MeshRange &mesh_range, const LocalFunction &local_f
...
 }
 //=================================================================================================//
+template <typename LocalFunction, typename... Args>
+void mesh_split_parallel_for(const MeshRange &mesh_range, const Arrayi &stride, const LocalFunction &local_function, Args &&...args)
+{
+    // forward sweeping
+    for (int m = 0; m < stride[0]; m++)
+        for (int n = 0; n < stride[1]; n++)
+        {
+            parallel_for(
```
Comment (WeiyiVirtonomy) [inline src/for_2D_build/meshes/mesh_iterators.hpp:94]: I haven't found `parallel_for(block, ...)` with a stride, so I used mod to skip the indices.
Alternatively, I can use two 1D loops `parallel(index begin, index end, index step, ...)` to do this. Which one do you prefer?

File: src/shared/meshes/cell_linked_list.hpp [hunk 2]
```diff
@@ -44,4 +44,20 @@ void CellLinkedList::searchNeighborsByParticles(
                  });
 }
 //=================================================================================================//
+template <class LocalDynamicsFunction>
```
Comment (WeiyiVirtonomy) [inline src/shared/meshes/cell_linked_list.hpp:47]: I set this function as a member function for now, but I can also move this part to `InteractionSplit` if you think that is better

File: src/shared/particle_dynamics/particle_dynamics_algorithms.h [hunk 3]
```diff
@@ -198,16 +199,15 @@ class InteractionSplit : public BaseInteractionDynamics<LocalDynamicsType, Paral
 {
   protected:
     RealBody &real_body_;
-    SplitCellLists &split_cell_lists_;
+    CellLinkedList &cell_linked_list_;
```
Comment (WeiyiVirtonomy) [inline src/shared/particle_dynamics/particle_dynamics_algorithms.h:202]: I can't find a way to make this function applicable to `BaseCellListList`, might need to write a new `SplitInteraction` function for `MultiLevelCellLinkedList`

**Comment by WeiyiVirtonomy** [comment]
I had some misunderstanding about the method, will fix it tomorrow

File: src/for_2D_build/meshes/mesh_iterators.hpp (hunk 1 above)
Comment (WeiyiVirtonomy) [reply src/for_2D_build/meshes/mesh_iterators.hpp]: Use two 1d `parallel_for` for now

File: src/shared/meshes/cell_linked_list.hpp (hunk 2 above)
Comment (WeiyiVirtonomy) [reply src/shared/meshes/cell_linked_list.hpp:47]: Keep as a member function of `CellLinkedList` for now

**Comment by WeiyiVirtonomy** [comment]
I think this PR is ready for review now

**Comment by WeiyiVirtonomy** [comment]
@DongWuTUM The computational time of damping is too long. I ran the same simulation without and with damping, the previous one took 104s and the later one took 199s, nearly a 90% increment. I'll try with the block_2d loop + mod and see if that one performs better

**Comment by Xiangyu-Hu** [comment]
@WeiyiVirtonomy I guess that there is something wrong.

File: src/shared/meshes/cell_linked_list.hpp [hunk 4]
```diff
@@ -44,4 +44,108 @@ void CellLinkedList::searchNeighborsByParticles(
...
+void CellLinkedList::particle_for_split(const execution::SequencedPolicy &, const LocalDynamicsFunction &local_dynamics_function)
+{
+    // foward sweeping
+    for (size_t k = 0; k < number_of_split_cell_lists_; k++)
+    {
+        // get the corresponding 2D/3D split cell index (m, n)
+        // e.g., for k = 0, split_cell_index = (0,0), for k = 3, split_cell_index = (1,0), etc.
+        const Arrayi split_cell_index = transfer1DtoMeshIndex(3 * Arrayi::Ones(), k);
```
Comment (WeiyiVirtonomy) [inline src/shared/meshes/cell_linked_list.hpp:55]: I have transformed the method to a 1D loop, but recomputing the corresponding cell indices every time can increase the computational cost a bit.
@DongWuTUM

Comment (Xiangyu-Hu) [reply src/shared/meshes/cell_linked_list.hpp:55]: @WeiyiVirtonomy how about the overall speed?

Comment (WeiyiVirtonomy) [reply src/shared/meshes/cell_linked_list.hpp:55]: The speed looks close to the old method. For test_3d_arch, the computational time on my machine is:
Old method: without damping 26.1719s, with damping 37.9359s, damping time 11.764s
New method: without damping 29.2173s, with damping 42.0216s, damping time 12.8043s

Comment (Xiangyu-Hu) [reply src/shared/meshes/cell_linked_list.hpp:55]: It looks good.

**Comment by WeiyiVirtonomy** [comment]
> @WeiyiVirtonomy As the master has been updated, the conflicts must be resolved first for the merge. i also noticed that there are some function is base mesh have repeating implementation, please also clean these repeating if there is any.

Merge conflict is now solved

File: src/shared/meshes/cell_linked_list.hpp [hunk 5]
```diff
@@ -96,7 +105,7 @@ void CellLinkedList::particle_for_split(const execution::SequencedPolicy &, cons
         // i_max = (M - m - 1) / 3 + 1, j_max = (N - n - 1) / 3 + 1
         // e.g. all_cells = (M,N) = (6, 9), (m, n) = (1, 1), then i_max = 2, j_max = 3
         const Arrayi all_cells_k = (all_cells_ - split_cell_index - Arrayi::Ones()) / 3 + Arrayi::Ones();
-        const size_t number_of_cells = get1DMeshSize(all_cells_k); // i_max * j_max
+        const size_t number_of_cells = get_1d_number_of_cells(all_cells_k); // i_max * j_max
```
Comment (Xiangyu-Hu) [inline src/shared/meshes/cell_linked_list.hpp:108]: Seems to use "all_cell_k.prod()" better?
https://eigen.tuxfamily.org/dox/group__TutorialReductionsVisitorsBroadcasting.html

Comment (WeiyiVirtonomy) [reply src/shared/meshes/cell_linked_list.hpp]: Thank you for the suggestion! I have replaced the function with `prod()`

**Review by Xiangyu-Hu — APPROVED** [review]

**Pull request merged**